src/main/java/com/denizenscript/denizencore/scripts/commands/file/YamlCommand.java

```diff
@@ -439,6 +444,7 @@ public void run() {
                                     }
                                     writer.write(outp);
                                     writer.close();
+                                    yaml.setDirty(false);

```

Comment:
This should go *before* the runnable (ie on line 432). The data is saved by the `saveToString`. The runnable is an async transfer of that saved data to the file-on-disk.

Reply:
saveToString is called by the to_text tag so didn't want to do anything in that method - as for putting this before the runnable, wouldn't that mean IOExceptions would incorrectly mark the object as not having changes? any way to get around this, or should we call setDirty(false) regardless of any write errors?

Reply:
No, I mean `setDirty` goes on line `432`. Before the runnable instead of inside it. I did not mean it goes in `saveToString`, of course it doesn't. I just meant that method is where the data that will be on file is finalized, and so setDirty should follow it (since that method is called on line 431, the setDirty thusly goes on 432).

---

Exceptions are exceptions. If write to disk fails... that should be treated as an error and responded to as such. has_changes will be inaccurate... but, well, who knows how much is wrong at that point - once you have internal errors, you got errors. As a general point of principle, "what if there's an IO exception" should be treated the same as "what if there's a RAM [soft error](https://en.wikipedia.org/wiki/Soft_error)"... errors are errors, there effects are outside our capacity to reliably deal with.

While you could technically just shove a `setDirty(true)` in the exception `catch` block... it's more proper to assume that if a write error happens, data consistency is forfeit regardless, and the server owner has some work to do (backup restoration, disk clearing, permissions/settings corrections+restarts, or something in that range). I'd advise against trying to handle exceptions like that at all, but if you choose to do so anyway I won't stop you.

---

The reason it matters to be *before* the runnable, if unclear: Imagine this situation:
- use `~yaml save` on a slow disk.
- a separate queue does `- yaml set x`
- *then* the `save` finishes, and does `setDirty(false)`.
In effect: the `x` set is applied in RAM, but not on file, but `has_changes` is false.

That's not an error case, it's perfectly reasonable to happen.
Essentially, being in the runnable causes a classic [race condition](https://en.wikipedia.org/wiki/Race_condition). Which, y'know, is bad.
Being before the runnable makes it be handled synchronously and thus no uncertainly, setDirty will always be correct (excluding any possible exceptions).

Reply:
Makes sense, thanks for the explanation